ldk/go/filesystemClient.go

```diff
@@ -156,16 +132,401 @@ func (f *FilesystemClient) ListenFile(ctx context.Context, path string, handler
 				continue
 			}
 
-			fi := FileInfo{
-				Name:    file.GetName(),
-				Size:    int(file.GetSize()),
-				Mode:    int(file.GetMode()),
-				Updated: t,
-				IsDir:   file.GetIsDir(),
+			fi := &FileInfo{
+				name:    file.GetName(),
+				size:    int(file.GetSize()),
+				mode:    int(file.GetMode()),
+				updated: t,
+				isDir:   file.GetIsDir(),
 			}
 			handler(FileEvent{Info: fi, Action: protoActionToAction(resp.GetAction())}, err)
 		}
 	}()
 
 	return nil
 }
+
+type GRPCFile struct {
+	ctx       context.Context
+	stream    proto.Filesystem_FilesystemFileStreamClient
+	fileMutex sync.Mutex
+	buffer    bytes.Buffer
+	readIndex int64
+}
+
+func newGRPCFile(ctx context.Context, stream proto.Filesystem_FilesystemFileStreamClient) *GRPCFile {
+	GRPCFile := &GRPCFile{
+		ctx:    ctx,
+		stream: stream,
+	}
+
+	return GRPCFile
+}
+
+func (f *GRPCFile) makeRequest() (int, error) {
+	err := f.stream.Send(&proto.FilesystemFileStreamRequest{
+		RequestOneOf: &proto.FilesystemFileStreamRequest_Read_{
+			Read: &proto.FilesystemFileStreamRequest_Read{},
+		},
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	for {
+		select {
+		case <-f.ctx.Done():
+			return 0, errors.New("context closed")
+		default:
+			resp, err := f.stream.Recv()
+			if err != nil {
+				return 0, err
+			}
+			switch resp.ResponseOneOf.(type) {
+			case *proto.FilesystemFileStreamResponse_Read_:
+
+				recievedBytes := resp.GetRead().GetData()
+				n := len(recievedBytes)
+				f.buffer.Write(recievedBytes)
+
+				if resp.GetRead().Error == "EOF" {
+					return 0, io.EOF
+				}
+
+				if resp.GetRead().GetError() == "" {
+					return n, nil
+				}
+				return n, errors.New(resp.GetRead().GetError())
+			default:
+				return 0, errors.New("unexpected response from server")
+			}
+
+		}
+	}
+
+}
+
+func (f *GRPCFile) Read(b []byte) (int, error) {
```

Comment:
I am likely calling this the wrong way, but I am getting `panic: runtime error: slice bounds out of range [:512] with capacity 0` inside this func.  [This is our Loop code](https://bitbucket.org/crosschx/loop-memorialcare-navigator/branch/SIDE-1035-memorial-care-loop-must-use-part2?dest=master#Lloop/loop.goT82).

What is the appropriate way to call this func?  I'm trying to go from `sidekick.Filesystem().Open()` which returns `ldk.File` to the actual file contents.

Reply:
@DrewRepaskyOlive Don't do the thing you are doing where you read it to a byte array. The `ldk.File` type is an `io.Reader`. So you can just pass it directly into `excelize.OpenReader` (https://godoc.org/github.com/360EntSecGroup-Skylar/excelize#OpenReader).

Reply:
Thanks @BillyMurrayOlive, that worked!